cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/TokenRangeMapping.java

```diff
@@ -22,55 +22,131 @@
 import java.io.Serializable;
 import java.math.BigInteger;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Function;
+import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ArrayListMultimap;
 import com.google.common.collect.Multimap;
 import com.google.common.collect.Range;
 import com.google.common.collect.RangeMap;
 import com.google.common.collect.TreeRangeMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import o.a.c.sidecar.client.shaded.common.response.TokenRangeReplicasResponse;
+import o.a.c.sidecar.client.shaded.common.response.TokenRangeReplicasResponse.ReplicaInfo;
 import o.a.c.sidecar.client.shaded.common.response.TokenRangeReplicasResponse.ReplicaMetadata;
-import org.apache.cassandra.spark.bulkwriter.RingInstance;
 import org.apache.cassandra.spark.common.model.CassandraInstance;
 import org.apache.cassandra.spark.data.ReplicationFactor;
 import org.apache.cassandra.spark.data.partitioner.Partitioner;
 import org.jetbrains.annotations.Nullable;
 
-// TODO: refactor to improve the return types of methods to use `Instance` instead of String and cleanup
-public class TokenRangeMapping<Instance extends CassandraInstance> implements Serializable
+public class TokenRangeMapping<I extends CassandraInstance> implements Serializable
 {
     private static final long serialVersionUID = -7284933683815811160L;
+    private static final Logger LOGGER = LoggerFactory.getLogger(TokenRangeMapping.class);
+
     private final Partitioner partitioner;
     private final ReplicationFactor replicationFactor;
-    private final transient Set<RingInstance> replacementInstances;
-    private final transient RangeMap<BigInteger, List<Instance>> replicasByTokenRange;
-    private final transient Multimap<Instance, Range<BigInteger>> tokenRangeMap;
-    private final transient Map<String, Set<String>> writeReplicasByDC;
-    private final transient Map<String, Set<String>> pendingReplicasByDC;
-    private final transient List<ReplicaMetadata> replicaMetadata;
+    private final transient Set<I> allInstances;
+    private final transient RangeMap<BigInteger, List<I>> replicasByTokenRange;
+    private final transient Multimap<I, Range<BigInteger>> tokenRangeMap;
+    private final transient Map<String, Set<I>> writeReplicasByDC;
+    private final transient Map<String, Set<I>> pendingReplicasByDC;
```

Comment:
I don't think it's worth keeping the replicas by DC, we always end up flattening. Why not just keep the set instead?